safe/impact_functions/inundation/flood_raster_road/impact_function.py

```diff
@@ -0,0 +1,474 @@
+            x0 = raster_xmin+(x*cell_width)
+            x1 = raster_xmin+((x+1)*cell_width)
+            y0 = raster_ymax-(y*cell_height)
+            y1 = raster_ymax-((y+1)*cell_height)
+            outer_ring = [
+                QgsPoint(x0, y0), QgsPoint(x1, y0),
+                QgsPoint(x1, y1), QgsPoint(x0, y1),
+                QgsPoint(x0, y0)]
+            # noinspection PyCallByClass
+            geometry = QgsGeometry.fromPolygon([outer_ring])
+            geometry.transform(ct)
+            f = QgsFeature()
+            f.setGeometry(geometry)
+            features.append(f)
+
+    _, features = vl.dataProvider().addFeatures(features)
+
+    # construct a temporary map for fast access to features by their IDs
+    # (we will be getting feature IDs from spatial index)
+    flood_cells_map = {}
+    for f in features:
+        flood_cells_map[f.id()] = f
+
+    # build a spatial index so we can quickly identify
+    # flood cells overlapping roads
+    if QGis.QGIS_VERSION_INT >= 20800:
+        # woohoo we can use bulk insert (much faster)
+        index = QgsSpatialIndex(vl.getFeatures())
+    else:
+        index = QgsSpatialIndex()
+        for f in vl.getFeatures():
+            index.insertFeature(f)
+
+    return index, flood_cells_map
+
+
+def _add_output_feature(
+        features,
+        geometry,
+        is_flooded,
+        fields,
+        original_attributes,
+        target_field):
+    """ Utility function to construct road features from geometry.
+
+    Newly created features get the attributes from the original feature.
+
+
+    :param features: A collection of features that the new feature will
+        be added to.
+    :type features: list
+
+    :param geometry: The geometry for the new feature. If the geometry is
+        multi-part, it will be exploded into several single-part features.
+    :type geometry: QgsGeometry
+
+    :param is_flooded: Flag indicating whenther the feature should be marked
+        as flooded.
+    :type is_flooded: bool
+
+    :param fields: Fields that should be assigned to the new feature.
+    :type fields: list
+
+    :param original_attributes: Attributes for the feature before the new
+        target field (see below) is added.
+    :type original_attributes: list
+
+    :param target_field: Output field used to indicate if the road segment
+        is flooded.
+    :type target_field: QgsField
+
+    :returns: A collection of features with the new feature appended.
+    :rtype: list
+    """
+    geoms = geometry.asGeometryCollection() if geometry.isMultipart() else [
+        geometry]
+    for g in geoms:
+        f = QgsFeature(fields)
+        f.setGeometry(g)
+        for attr_no, attr_val in enumerate(original_attributes):
+            f.setAttribute(attr_no, attr_val)
+        f.setAttribute(target_field, is_flooded)
+        features.append(f)
+
+
+def _union_geometries(geoms):
+    """ Return a geometry which is union of the passed list of geometries """
```

Comment:
Sorry this one needs similar treatment too:

geoms -> geometries

And docstrings for parameters and return